docs/spec/concepts.rst

```diff
@@ -3,6 +3,268 @@
+the fully static type ``object`` is the set of all Python objects. The fully
+static type ``bool`` is the set of values ``{ True, False }``. The fully static
+type ``str`` is the set of all Python strings; more precisely, the set of all
+Python objects whose runtime type (``__class__`` attribute) is either ``str``
+or a class that inherits directly or indirectly from ``str``. A :ref:`Protocol
+<Protocols>` denotes the set of all objects which share a certain set of
+attributes and/or methods.
+
+Gradual types
+~~~~~~~~~~~~~
+
+:ref:`Any` represents an unknown static type. It denotes some unknown set of
+values.
+
+This may appear similar to the fully static type ``object``, which represents
+the set of all Python objects, but it is quite different.
+
+If an expression has the type ``object``, a static type checker should ensure
+that operations on the expression are valid for all Python objects, or else
+emit a static type error. This allows very few operations! For example, if
+``x`` is typed as ``object``, ``x.foo`` should be a static type error because
+not all Python objects have an attribute ``foo``.
+
+An expression typed as :ref:`Any`, on the other hand, should be assumed to have
+_some_ specific static type, but _which_ static type is not known. A static
+type checker should not emit static type errors on an expression or statement
+if :ref:`Any` might represent a static type which would avoid the error. (This
+intuition is made more precise below, in our definitions of materialization and
+assignability.)
+
+Similarly, a gradual type such as ``tuple[int, Any]`` (see :ref:`tuples`) or
+``int | Any`` (see :ref:`union-types`) does not represent a single set of
+Python objects; rather, it represents a (bounded) range of possible sets of
+values.
+
+In the same way that ``Any`` does not represent "the set of all Python objects"
+but rather "an unknown set of objects", ``tuple[int, Any]`` does not represent
+"the set of all length-two tuples whose first element is an integer." That is a
+fully static type, spelled ``tuple[int, object]``.  In contrast, ``tuple[int,
+Any]`` represents some unknown set of tuple values; it might be the set of all
+tuples of two integers, or the set of all tuples of an integer and a string, or
+some other set of tuple values.
+
+In practice, this difference is seen (for example) in the fact that we can
+assign an expression of type ``tuple[int, Any]`` to a target typed as
+``tuple[int, int]``, whereas assigning ``tuple[int, object]`` to ``tuple[int,
+int]`` is a static type error. (Again, we formalize this distinction in the
+below definitions of materialization and assignability.)
+
+In the same way that the fully static type ``object`` is the upper bound for
+the possible sets of values represented by ``Any``, the fully static type
+``tuple[int, object]`` is the upper bound for the possible sets of values
+represented by ``tuple[int, Any]``.
```

Comment:
There is an inherent tension here in wanting to use examples to clarify the concepts, while some of the examples themselves implicitly depend on more advanced concepts (like the idea of a generic type in the first place!) that we aren't ready to dive into yet.

That said, in this particular section I don't think there is any implicit assumption about variance. (There definitely is later on in the section on consistent subtyping.) I think everything said here would also be true of list?

Reply:
I'm not sure `list[object]` is the upper bound for the possible sets of values represented by `list[Any]`, is it? `list[object]` is not a supertype of `list[int]`, but `list[Any]` could be materialised as `list[int]`

Reply:
Regarding the case below where we do assume covariance of tuple, in discussing consistent subtyping: I think in practice this is OK because covariance is definitely most people's natural intuition ("yes, a list of strings is a list of objects!"). It's invariant generic containers that people find unintuitive and need explanation for. So I think it works fine to use tuples in examples in this doc, with the assumption of covariance, and leave further discussion of variance to the generics doc.

Reply:
Definitely agree that a detailed discussion of variance is out of scope here.

Reply:
Ah, of course you're right, `list[object]` is not an upper bound for `list[Any]`, exactly as you say. There is a covariance assumption here as well.